lib/api/core/storage/clientAdapter.js

```diff
@@ -0,0 +1,165 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const
+  errorsManager = require('../../../config/error-codes/throw')
+    .wrap('internal', 'storage_engine');
+
+/**
+ * Storage client adapter to perform validation on index/collection existence
+ * and to maintain the index/collection cache.
+ */
+class ClientAdapter {
+  /**
+   * @param {Elasticsearch} storageClient
+   * @param {IndexCache} indexCache
+   */
+  constructor (storageClient, indexCache) {
+    this._client = storageClient;
+    this._indexCache = indexCache;
+
+    // Methods that needs to assert index and collection existence
+    this._assertIndexAndCollectionMethods = [
+      'search',
+      'get',
+      'mGet',
+      'count',
+      'create',
+      'createOrReplace',
+      'update',
+      'replace',
+      'delete',
+      'deleteByQuery',
+      'getMapping',
+      'updateMapping',
+      'truncateCollection',
+      'import',
+      'refreshCollection',
+      'exists',
+      'mCreate',
+      'mCreateOrReplace',
+      'mUpdate',
+      'mReplace',
+      'mDelete'
+    ];
+
+    // Methods that needs to assert index existence
+    this._assertIndexMethods = [
+      'listCollections'
+    ];
+
+    // Methods directly bind to the storageClient
+    this._rawMethods = [
+      'scroll',
+      'listIndexes',
+      'listAliases',
+      'info',
+      'indexExists',
+      'collectionExists',
+      'init',
+      'createIndex'
+    ];
+
+    // Methods that need to assert index/collection existence and add/remove item
+    // from the indexCache
+    this._othersMethods = [
+      'createIndex',
+      'createCollection',
+      'deleteIndexes',
+      'deleteIndex'
+    ];
+
+    for (const method of this._assertIndexAndCollectionMethods) {
+      this[method] = (index, collection, ...args) => {
+        this._assertIndexAndCollectionExists(index, collection);
+
+        return this._client[method](index, collection, ...args);
+      };
+    }
+
+    for (const method of this._assertIndexMethods) {
+      this[method] = (index, ...args) => {
+        this._assertIndexExists(index);
+
+        return this._client[method](index, ...args);
+      };
+    }
+
+    for (const method of this._rawMethods) {
+      this[method] = (...args) => this._client[method](...args);
+    }
+  }
+
+  async createCollection (index, collection, mappings) {
+    const response = await this._client.createCollection(index, collection, mappings);
+
+    this._indexCache.add({ index, collection, scope: this._client.scope });
+
+    return response;
+  }
+
+  async deleteIndex (index) {
+    this._assertIndexExists(index);
+
+    const response = await this._client.deleteIndex(index);
+
+    return response;
+  }
+
+  async deleteIndexes (indexes) {
+    for (const index of indexes) {
+      this._assertIndexExists(index);
+    }
+
+    const response = await this._client.deleteIndexes(indexes);
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Immediately return this expression instead of assigning it to the temporary variable "response". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3AS1488)